src/libcephsqlite.cc

```diff
@@ -0,0 +1,701 @@
+    _cephsqlite3_Access,             /* xAccess */
+    _cephsqlite3_FullPathname,       /* xFullPathname */
+    _cephsqlite3_DlOpen,             /* xDlOpen */
+    _cephsqlite3_DlError,            /* xDlError */
+    _cephsqlite3_DlSym,              /* xDlSym */
+    _cephsqlite3_DlClose,            /* xDlClose */
+    _cephsqlite3_Randomness,         /* xRandomness */
+    _cephsqlite3_Sleep,              /* xSleep */
+    _cephsqlite3_CurrentTime,        /* xCurrentTime */
+  };
+  // std::cerr << std::hex << pthread_self() << ":" << __FUNCTION__ << "::" << std::endl;
+  return &_cephsqlite3_vfs;
+}
+
+extern "C"
+void __attribute__ ((constructor)) _cephsqlite3__vfs_register()
+{
+  // std::cerr << std::hex << pthread_self() << ":" << __FUNCTION__ << "::" << std::endl;
+  sqlite3_vfs_register(_cephsqlite3__vfs(), 1);
+}
+
+extern "C"
+void __attribute__ ((destructor)) _cephsqlite3__vfs_unregister()
+{
+  // std::cerr << std::hex << pthread_self() << ":" << __FUNCTION__ << "::" << std::endl;
+  sqlite3_vfs_unregister(_cephsqlite3__vfs());
+}
+
+extern "C"
+sqlite3 *ceph_sqlite3_open(
+  librados::Rados &cluster,
+  const char *dbname,           /* eg. "kvstore" instead of "kvstore.db" */
+  const char *rados_namespace,
+  int ceph_pool_id,
+  int stripe_count,
+  int obj_size,
+  bool must_create
+)
+{
+  // std::cerr << std::hex << pthread_self() << ":" << __FUNCTION__ << "::" << std::endl;
+  int ret = -1;
+  /* FIXME
+   * how long should the io_ctx and cluster objects exist for radosstriper
+   * to be functional ?
+   */
+  librados::IoCtx *io_ctx = new librados::IoCtx;
+  libradosstriper::RadosStriper *rs = new libradosstriper::RadosStriper;
+
+  ret = cluster.ioctx_create2(ceph_pool_id, *io_ctx);
+  if (ret < 0) {
+    /* unable to create the IO Context */
+    return NULL;
+  }
+
+  io_ctx->set_namespace(rados_namespace);
+
+  ret = libradosstriper::RadosStriper::striper_create(*io_ctx, rs);
+  if (ret < 0) {
+    /* unable to create the striper */
+    return NULL;
+  }
+
+  uint64_t alignment = 0;
+  ret = io_ctx->pool_required_alignment2(&alignment);
+  if (ret < 0) {
+    /* no alignment retrieved */
+    return NULL;
+  }
+
+  rs->set_object_layout_stripe_unit(alignment);
+  rs->set_object_layout_stripe_count(stripe_count);
+  rs->set_object_layout_object_size(obj_size);
+
+  sqlite3 *db = NULL;
+  const int db_open_flags = SQLITE_OPEN_NOMUTEX       | /* single client access */
+                            SQLITE_OPEN_PRIVATECACHE  |
+                            SQLITE_OPEN_READWRITE     |
+                            SQLITE_OPEN_URI           |
+                            (must_create ? SQLITE_OPEN_CREATE : 0);
+
+  std::stringstream ss;
+
+  /* pass the address of the RadosStriper C++ object in the URI so that the VFS
+   * methods can access it
+   */
+  std::string mode = (must_create ? "rwc" : "rw");
+  /* create a URI based file name */
+  ss << "file:" << dbname << ".db?mode=" << mode << "&cache=private&vfs=cephsqlite3";
```

Comment:
consider I may have multiple sqlite3 databases named "dbname" in a pool. The only way to distinguish is the rados namespace. You also have the same problem for different pools.

I don't see a particular reason to make this file name particularly user friendly. Can we just do:

`"libcephsqlite:<pointer to internal CephVFSContext>"`

?

Reply:
If we create a database by the convention you've mentioned above, it'll be difficult to locate it after the system restarts since the pointer value will be lost. Or am I lost here ?

I was under the impression that the client apps i.e. ceph daemons or some other third-party user applications, should be able to use the library and locate the database without issues, since the database will be persistent. Hence the user-friendly database naming feature.

Or, I may not have understood the intended use case here.

Reply:
> @batrick If we create a database by the convention you've mentioned above, it'll be difficult to locate it after the system restarts since the pointer value will be lost. Or am I lost here ?

If the application restarts then it will use the `ceph_sqlite3_open` function again to reopen it. The application should never see or care what the sqlite3 URI contains.

As for my suggested format, it would be best to not use a pointer actually. Simpler would be just some uuid token to let you lookup the pointer.

Reply:
@batrick The application is only supposed to pass the name of the database eg. "accounts", and only use that name to access the database every single time.
The URI encoding is done by the`ceph_sqlite3_open()` implementation and the entire URI name does not persist on the disk. If the command-line tools are used to list the objects in the pool and namespace, then user will only get to see "accounts.db.0000000000000000", "accounts.db-journal"  and some other similarly named files. All the other URI parameters are parsed by sqlite and held in memory.

I didn't quite understand how the user/client app will get to know the UUID of the database beforehand unless that's whats passed to `ceph_sqlite3_open()`.

Reply:
> I didn't quite understand how the user/client app will get to know the UUID of the database beforehand unless that's whats passed to ceph_sqlite3_open().

You misunderstood me. I mean that libcephsqlite internally will use a uuid string to associate with the open database pointer (i.e. all the contextual state with the open libradosstriper connections). 

But now I'm not really sure I understand the interface correctly. I will need to try out your PR first before giving you more feedback on this matter.